### ISSUE TITLE
New UI

### ISSUE THREAD

**Issue by meandor** [ampache/ampache#359]
So I had an issue that got me thinking over at my repo (https://github.com/meandor/ampdroid/issues/37) about the whole development of Ampache right now.

When I first installed Ampache I hated the design and I thought the whole structure is so complicated and outdated and difficult to use. I installed the penguin theme thingy and it was a tiny bit better,

The whole UI got changed with this reborn theme and I actually like it more now but it still is overloaded with a bunch of stuff that I don't really get why it does even exist sometimes.
It is still not so easy to use and not really intuitive. We still kinda have the frame layout.

So as I mentioned before this was inspired by the issue at my repo where I stated that I kinda have the feeling that Ampache is getting more and more bloated with stuff that is not sooo important and the main development is getting a bit distracted. I would wish for the user expereince and Ampache to get simpler, easier to use and an overall perfect music streaming experience.

I could not let this feeling and idea go and I made a quick mockup of how I would wish the user experience of Ampache to be.
The one thing I love about open source software is the collaboration of people who basically want the same thing and try to make what they want and inspire and influence each other which often leads to great results.

So I hope with this I can maybe convince some people that the user experience could be much better and easier and simpler and maybe even make people who have the time actually work with this/develop this further.

I hope I didn't step on anybodys toes from the developing team.

![ampache-ui-home](https://cloud.githubusercontent.com/assets/4395138/3758028/074ea250-1841-11e4-9737-309dc359a80f.png)
![ampache-ui-music](https://cloud.githubusercontent.com/assets/4395138/3758029/07539c06-1841-11e4-8244-45e2d14e3f9f.png)

**Comment by pserafini** [comment]
Hi guys, really looking forward to testing this new theme.

**Comment by a7medo778** [comment]
looking forward for this guys, any updates

**Comment by a7medo778** [comment]
i was actually trying to do my theme for the community, i hired a dev to link the api's to this angularJS template
http://themeforest.net/item/groovy-angularjs-music-app-template/11426974

but he told me there was an issue with how the script it self is designed, anyway he guided me to a solution and i thought i should put it here if any one was going to do it
"

Well while it sounds like a good option it will still require quite a bit of development. also the API returns XML documents, not the ideal at all for a framework like angular or any good front end framework. You can see a discussion here : https://github.com/ampache/ampache/issues/449

This requires also some knowledge of this "susonic?" you mention and the whole ampache stack.

Unfortunately I don't think I have the time for such a project but I am happy to just point in the right direction on how I think it would work if you go for the theme. 

Basically angular services calling the API services of ampache, (ampache does provide more things than the theme has maybe) but you might need someone with knowledge of both Javascript and the ampache stack to be able to adjust whichever is necessary. 

The theme right now like I mentioned just has some json files simulating what the server is with artists, albums, songs etc.. and already has the HTTP Requests for these files, basically necessary would be to replace these calls with calls to the ampache API
"

**Comment by SUTJael** [comment]
Souds great @a7medo778, but of course you will need some API improvements to have something working properly. Don't hesite to open new tickets with required improvements. Unfortunately it's not possible for us to improve the native ampache UI with that kind of components (too much constraints due to the coding style). Please keep us informed when you have something working.
I do not have much time to work on Ampache since few months, but I hope I will be able to continue working on the new UI in coming months ! :)

**Comment by zaphbbrox** [comment]
Any news? And great work BTW

**Comment by SUTJael** [comment]
No work has been done on that sorry... Hoping one day I will have time to work on that again :)

**Comment by Psy-Virus** [comment]
Too bad. :'(
But @SUTJael, does this mean I finally have the "GO!" for [this](https://groups.google.com/forum/#!searchin/ampache/translatable/ampache/3v4uEXVTf6U/1zDjjVv1C7IJ)?
`<sarcasm>`I only wait for an answer since April but who cares!`</sarcasm>` :stuck_out_tongue_winking_eye:
But I didn't ask again, so it's also my fault.

The other translators will kill me if I review/rewrite everything after there was so much progress. :cold_sweat: 
But from my point of view, it has to be done.

:-)

**Comment by Afterster** [comment]
New UI would be great but according to the progress made on it, no one should wait for this new UI before making architectural changes or it would put the project on a dead state. I don't :smile:.

**Comment by Psy-Virus** [comment]
Of course, but he said I shouldn't do anything until the release of the new UI, so I didn't. ^^
Well then I'll begin except one of you both see any problem with it.
@Afterster, would it be a problem, if I create a dummy PR that I'll update every time I do something and for discussion and maybe if I need help and advice? It will contain one commit, because I would pull Ampache one time and force-update the repository maybe after every work run. I also won't update this repository so it will remain on the pulled version. It would be to much work to go with every new update and merge it.

The rewriting of the German translation took me around 6 hours power work :grin: , I hope to get this task completed within a week or two (maybe with completed German translation), so that there are not that many conflicts at the end.

What do you think?

:-)

**Comment by Afterster** [comment]
Well you can create the PR for discussion and update it, but please try to keep a clean git history. My time is also limited, I will not review your PR on a daily frequency so don't hesitate 'to have something' before creating the PR. Thanks!

**Comment by Psy-Virus** [comment]
@Afterster, we definitely need to work on our communication. :D
As I said before, The PR will only contain ONE commit, because I will force the update of my repo (it gets completely overwritten) so the created Commit will get updated and you have to review only one commit what contains every FINAL change after I'm ready.
So don't worry, I let you know if it's ready, it will be the last comment that will contain a not to misunderstand "It's finished or equal" :-).
You only would have to take a look, if I please you; like if I need help and no one else where able to help me but this also would more run via discussion in the comments. So I think you only need to take a look at it, if I am finished. :-)

I hope this cleared it up. And sorry, if it sounded a little bit harsh, it definitely wasn't meant this way :-)

All great this way? Or do you mean anything else with "keeping the history clean"? As I understood it in the past it's "the more commits, the uglier the mergin work. because of udates from updates and so on"

**Comment by Afterster** [comment]
Ok! :smile: 
Thanks.

**Comment by SUTJael** [comment]
I'll have a 2-days AngularJS training next year. Hoping it will get me some basic ideas if this is something we could introduce in Ampache. I think making the new UI would be the perfect time to review the architecture too (otherwise it will never be done).
When started the new UI, that was really hard to change all templates but keeping the ugly controllers behind.
We could also have a look to NodeJS/BackboneJS, but we need some experts :)

**Comment by Afterster** [comment]
Ok everybody... good news are coming.

I'm currently moving Ampache to Laravel framework (5.2), I will try to keep all features in the process but few will be dropped because it is simply bad practice. Anyway, nothing will be published before weeks/months until I get something working (current code architecture is a nightmare to have something running on Laravel). Because almost everything have to be rewritten, this version will be a major version (Ampache 4) with breaking changes ; only APIs will remains. Ampache 3.x will remains in parallel to provide bug fix releases.
But, I'm posting this message here for those willing to have a new design. I'm migrating current Reborn theme to Laravel views, based on Blade template engine. This way the view would be easily changed. I know @Niols is waiting for this, but @meandor, @SUTJael don't hesitate to prepare independent blade templates on your side with your new UI trial. I believe it could easily be done this way.
About Angular.JS, we should be able to have both kind of view on the same installation according to theme choice. All controllers will provide templated view by default, but also JSON RESTful content if Ajax request that could be consumed by Angular.JS. Main work will be to create a common JavaScript controllers library, and then a first Angular.JS theme. This is a long term discussion but will be possible with the move to Laravel.

**Comment by arnaudbey** [comment]
Ouch, what a (nice) change ! Moving to a framework is a big work, but will help a lot (i hope). Good luck !

**Comment by SUTJael** [comment]
Thanks @Afterster for the update. It's quite a synchro: I was thinking during last weeks of what we could use for a major rewrite.
I started this weekend to work on Meteor/React which seems to be the top winning techno in coming years. For the moment I'm just doing some tutorials to test and learn the framework but it seems very promising. At the same time, MongoDB seems to be perfect for Ampache.

What do you think about that? Wouldn't be the good time ?

**Comment by a7medo778** [comment]
awesome news :dancer: 
now just to be sure, i was working with https://www.oddnetworks.com/ to make their platform work with ampache, that shouldn't effect what we are doing right ?

as for the upgrade it self, i'll start reading about laravel since i really don't know what it is and its history :P 
i know meteor have a very active community, mongoDB is diffenatly worthy of considering for db, and while at it would be nice if scaling was taken into consideration 

but in general, awesome news indeed :D but we want a killer UI for the uploaders to work with

**Comment by Afterster** [comment]
Meteor is Node.JS only, even if it is a major rewriting I don't want to change target technology, we should stay on PHP. I believe you're talking about MongoDB because of Meteor.
Laravel provides a nice ORM (Eloquent) which is really working and provide database engine abstraction for MySQL, PostgreSQL and SQLite. If you're stuck to MongoDB idea, someone made a MongoDB extension for Laravel with Eloquent support, see https://github.com/jenssegers/laravel-mongodb but should be tested first.

@a7medo778 I believe oddworks will use APIs so it should be fine about the compatibility. It will only be broken if they are directly changing PHP code or accessing database but I don't think so.

**Comment by a7medo778** [comment]
I been doing some reading, Granted that I never seen one with 5.2, but laravel always seems to be the slowest, lowest throughput and highest in memory consumption compared to other php frameworks benchmarks
Am I missing any thing ?

Regards
Ahmed

**Comment by SUTJael** [comment]
By staying on PHP, Laravel is sure the better solution I totally agree. Migrate to Meteor was, from what I saw, the best possible choice but that would of course add many listen/rewrite efforts. And as you're the one which really dev right know, your choice is understandable :)
That said, I will definitely review the blade's views, was to painful to do the new ui with the current archi/source code. Don't hesitate to tell me when you've something I can play with.
I think that's also the best time to review the database as we were missing some important abstraction levels (like "disks" for exemple).

**Comment by wagnered** [comment]
@Afterster,

If I may make a suggestion, assuming you have not already thought of it,
 I suggest that as many features as can be added be created as packages
in which users can choose to add through composer.  There are many
features in 3.8.2 that I have no use for and would like to ignore.





On 03/13/2016 11:21 PM, Afterster wrote:
> Ok everybody... good news are coming.
> 
> I'm currently moving Ampache to Laravel framework (5.2), I will try to
> keep all features in the process but few will be dropped because it is
> simply bad practice. Anyway, nothing will be published before
> weeks/months until I get something working (current code architecture is
> a nightmare to have something running on Laravel). Because almost
> everything have to be rewritten, this version will be a major version
> (Ampache 4) with breaking changes ; only APIs will remains. Ampache 3.x
> will remains in parallel to provide bug fix releases.
> But, I'm posting this message here for those willing to have a new
> design. I'm migrating current Reborn theme to Laravel views, based on
> Blade template engine. This way the view would be easily changed. I know
> @Niols <https://github.com/Niols> is waiting for this, but @meandor
> <https://github.com/meandor>, @SUTJael <https://github.com/SUTJael>
> don't hesitate to prepare independent blade templates on your side with
> your new UI trial. I believe it could easily be done this way.
> About Angular.JS, we should be able to have both kind of view on the
> same installation according to theme choice. All controllers will
> provide templated view by default, but also JSON RESTful content if Ajax
> request that could be consumed by Angular.JS. Main work will be to
> create a common JavaScript controllers library, and then a first
> Angular.JS theme. This is a long term discussion but will be possible
> with the move to Laravel.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ampache/ampache/issues/359#issuecomment-196164106>.
> 

-- 
73 From
Ernie D
Federal Way, WA | CN87ug

**Comment by Niols** [comment]
I guess that is the aim in the end.
But Ampache isn't modular enough (yet) to allow such features to be separated from the main branch.

I guess there could be packages `ampache-ldap`, `ampache-oauth`, `ampache-dropbox`, …
But then one would need an easy way to install such packages. I'm not sure every user would want to handle `composer` by himself. Maybe a composer interface in Ampache?

@Afterster I already told you I found this project very exciting. Let me know if there's anything I can do to help you in this huge refactor

**Comment by Razrael** [comment]
Hey there

Its a pity I didn't get notified here, because I was working on similar stuff and just missed the conversation here.

Last year I promised you to work on the Database Abstraction. So I was working at it quite a bit in the past and could finish it mostly this weekend.

I also played around with implementing a new templating engine (Fluid from TYPO3 / Flow) and doing the whole stuff in a better MVC manner. Also I rewrote the bootstrapping part...

Please take a look at [my work](https://github.com/ampache/ampache/compare/develop...Razrael:feature/addPersistanceLayer), especially the views in /Resources and the possibilities with the new Database Layer, [showcased here](https://github.com/ampache/ampache/compare/develop...Razrael:feature/addPersistanceLayer#diff-f1afaa3649e8200a21c06f80441b5870R36)

If you check out my branch, be warned: There is currently no UI, just showcase. So it will just print out a list of songs, which might not work if you have really many in there.

And now I saw this conversation here. I hope I didn't do the whole work for nothing.
@Afterster Did you already start with the rewrite? It would be good if you could involve us, because so much work could just get lost.

**Comment by Afterster** [comment]
@wagnered @Niols this is indeed the goal. Most dependencies should be optional today as it is only requires if using a specific feature.

@Razrael oh :smile:! I started some work on Laravel already (UI migrated, some basic features...) but this will takes some additional months (on the road for almost 1 month btw, lake of activity has a reason ;)). It will be difficult / time consuming to do the migration publicly with several contributors as long the basics are not ready. That's my opinion but any help will be really appreciated afterward. Not sure it's the perfect solution, your work is also interesting (thanks!) and could be a temporary solution ; sometimes temporary solution remains... ; but Ampache really need a framework to be more efficient / have a better code base.

**Comment by Phyks** [comment]
Just an idea I had while browsing some issues about UI (and found this really interesting one that I'll be following up) is to offer a "keyboard only experience" in Ampache as well. I would love to be able to use my keyboard to navigate easily through Ampache, add songs to playlist, edit playlist, find songs etc.

**Comment by nasariasl** [comment]
hey guys :+1: 
this new ui is mobile friendly ?

**Comment by Phyks** [comment]
Hi,

Just so you know I am actually giving it a try, to get a new interface for Ampache. The result is [here](https://github.com/Phyks/ampache_react) (but is a big WIP).

I did not write it as a new default interface for the moment, rather as an alternative interface, to live side by side with Ampache and eventually get integrated in core in the end.

For this, it relies on Ampache XML API (but could be easily changed to whatever API providing the same features). It is built around React, which seems to be one of the most used library for this at the moment, and has cool features such as enabling easy server side rendering and compiling to native apps.

I mostly finish a first layout and setup of dev tools, and am going to work on further integration (fix of design bugs, webplayer etc). For now, it basically just allows you to browse through your music collection (no player yet, as it relies on discussion in another related issue).

If you give it a try and/or want to contribute in any way, please report any bugs you might find, and any UX idea you might have! Thanks :)

**Comment by a7medo778** [comment]
just added a bounty for the laravel efforts, maybe to get @wagnered a big mac for his hard work :P

**Comment by ronilaukkarinen** [comment]
Noticed the same, started using Ampache today and as a front end designer I'm not quite pleased with the UI overall. Functionality is brilliant though. So, I decided I'm going to solve this by making my own theme for my minimalistic needs, will take some time but eventually getting there... [I have already contributed to Subsonic](https://github.com/ronilaukkarinen/rolleshark) just like that even though its (older) UI is more messed up.

**Comment by wagnered** [comment]
Hello @ronilaukkarinen 

The (future) Ampache is being based on Laravel and w3.css and will have a more responsive  design.  I've been working mostly on the back end, creating an adequate front end to support designing the logic.  It's a reasonable facsimile of the original front end. :smile: And I'm designing the web pages only as needed at the time.  

I can't give a time line on a release as I'm the only active developer working on the front end and logic.

**Comment by ronilaukkarinen** [comment]
@wagnered OK, thanks for your answer, perhaps it's not wise then for me to use much time to improve the current version then. Will follow this repo and maybe help when the time is right. Thank you for your effort and as in open source in general, take your time :) sounds very promising!

**Comment by lachlan-00** [comment]
this branch of the ui won't be updated so this issue will be closed.

#2161 is the current "new" works on the ui.
